### ISSUE TITLE
distfile doesn't ship m4 directory

### ISSUE THREAD

**Issue by clausecker** [unfs3/unfs3#26]
The unfs3-10.0.0.tar.gz tarball doesn't ship the m4 directory.  Thus we are unable to patch the configure script in the FreeBSD port.  Please re-roll the distfile with the m4 directory.

I also noticed that you stopped shipping the pre-generated parser in 10.0.0.  While not needed for our purposes, you might be interested in shipping it for consistency.

**Comment by CendioOssman** [comment]
> The unfs3-10.0.0.tar.gz tarball doesn't ship the m4 directory. Thus we are unable to patch the configure script in the FreeBSD port. Please re-roll the distfile with the m4 directory.
> 

That should probably be included, yes. But it wasn't included before, so I don't quite understand why it is suddenly a requirement for you?

What modifications do you need to do? Perhaps it's something we can include directly?

> I also noticed that you stopped shipping the pre-generated parser in 10.0.0. While not needed for our purposes, you might be interested in shipping it for consistency.

That might have been accidental. The system isn't really designed with flex and yacc being optional, and the dist rule isn't supposed to include those files. IOW, it could be a nice feature, but it probably needs more work than just including the files.

**Comment by clausecker** [comment]
> That should probably be included, yes. But it wasn't included before, so I don't quite understand why it is suddenly a requirement for you?

The previous version 0.9.22 did not have an m4 directory if I recall correctly.

> What modifications do you need to do? Perhaps it's something we can include directly?

We are building unfs3 for FreeBSD.  You can see a list of patches we apply [in this directory](https://github.com/freebsd/freebsd-ports/tree/main/net/unfs3/files).  My [patch to update to 0.9.23](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=238161&action=diff) is currently pending with some patches being dropped and new patches being added.

Basically, we have to remove some of your configure tests as they don't work correctly on FreeBSD.   Then we need to regenerate the `configure` script which requires the `m4` directory.  I can provide more explanations on the individual patch items if desired.

We currently cannot upgrade to 10.0.0 as the hacked sunrpc implementation you use does not build on FreeBSD and does not seem to be trivial to port.  Due to changed signatures, the code does not build with standard sunrpc (as FreeBSD ships it) either.

**Comment by CendioOssman** [comment]
> The previous version 0.9.22 did not have an m4 directory if I recall correctly.

Ah, you're right. It should be fixed in 49b0d14e9e1b46005c4701203c6d0f0b69e0d44b. I've re-rolled some new tar balls.

> We are building unfs3 for FreeBSD. You can see a list of patches we apply [in this directory](https://github.com/freebsd/freebsd-ports/tree/main/net/unfs3/files). My [patch to update to 0.9.23](https://bugs.freebsd.org/bugzilla/attachment.cgi?id=238161&action=diff) is currently pending with some patches being dropped and new patches being added.

Those patches look fairly minor and should be something we could get in here upstream.

> We currently cannot upgrade to 10.0.0 as the hacked sunrpc implementation you use does not build on FreeBSD and does not seem to be trivial to port. Due to changed signatures, the code does not build with standard sunrpc (as FreeBSD ships it) either.

That's unfortunate. Have you informed libtirpc about this? We can't really stay with sunrpc as glibc has dropped that. :/

**Comment by clausecker** [comment]
Honestly I do not understand why the tirpc people decided to make their API incompatible with standard Sun IPC.  If they want to have IPv6 support, they should have done so with new function names so as to avoid collisions.  As is, it seems questionable if tirpc can be used safely on FreeBSD as it will interfere with any program components or libraries that use the standard Sun RPC.

I've actually tried to port tirpc to FreeBSD in order to port version 10.0.0, but there are lots of identifier conflicts with FreeBSD headers. Apparently large parts of tirpc have been ripped from BSD sources with no renaming.  I eventually gave up after the patches became tricky to manage.

As I understand it, it does not seem to be possible for you to design the software such that it compiles both with standard SUN RPC and tirpc.  Is that correct?

**Comment by CendioOssman** [comment]
> Honestly I do not understand why the tirpc people decided to make their API incompatible with standard Sun IPC. If they want to have IPv6 support, they should have done so with new function names so as to avoid collisions. As is, it seems questionable if tirpc can be used safely on FreeBSD as it will interfere with any program components or libraries that use the standard Sun RPC.

That's odd. I'm fairly sure libtirpc coexisted with glibc's sunrpc for quite some time.

> As I understand it, it does not seem to be possible for you to design the software such that it compiles both with standard SUN RPC and tirpc. Is that correct?

To be honest, we didn't really investigate that as the primary target was Linux and there was no sunrpc to even test with there.

**Comment by kraileth** [comment]
I've also tried to build version 0.10.0 on FreeBSD and of course failed as there's no `libtirpc` available. After doing some research, I eventually came here to look at the issues of this project.

Well, it seems that tirpc is highly Linux-specific and the only port to a non-Linux platform that I could find was for Haiku where a developer obviously succeeded in patching it enough to make it work. Looks like there is not a single port to any BSD and not to any illumos distribution, either (which is pretty ironic, since this means that the platform where NFS started is now getting cut off...).

I understand the need to move to tirpc on Linux, but this breaks unfs3 basically for the rest of the world which is rather unfortunate, especially when keeping in mind that ensuring portability helps projects even when they focus on Linux (which admittedly is a somewhat sensible thing to do today). Unfs3 is in high regards on FreeBSD as it allows for running an NFS server in a jail (which doesn't work with the base system NFS server). If you're just missing an environment to have a look, I can offer to give you SSH access to a development jail.

Unfs3 is currently packaged and used on FreeBSD, NetBSD, DragonFly BSD and macOS, in addition to Linux (or at least these are the ones that I know of). If no solution can be found, that would effectively mean killing off those platforms.

**Comment by CendioOssman** [comment]
We're starting to drift off-topic here. The original problem is resolved, so I'm going to close this issue.

Support for original sunrpc could certainly be considered, so feel free to open a new issue to continue that discussion. In the end it will come down to someone doing the work and getting a suggestion in place on how we can cleanly support both libtirpc and sunrpc.
